src/rust/generator-dispatcher/src/generator_ids_cache.rs

```diff
@@ -0,0 +1,232 @@
+                                            .get_generators_for_event_source(GetGeneratorsForEventSourceRequest {
+                                                event_source_id
+                                            })
+                                            .await;
+
+                                        drop(retry_client_guard); // release the client lock
+                                    };
+
+                                    result
+                                        .expect("fatal error, unknown state")
+                                        .plugin_ids
+                                },
+                            };
+
+                        generator_ids_cache.insert(
+                            event_source_id,
+                            generator_ids,
+                        ).await;
+                    }
+                })
+                .await;
+        });
+
+        Ok(Self {
+            generator_ids_cache: generator_ids_cache_retval,
+            updater_tx,
+        })
+    }
+
+    /// Retrieve the generator IDs corresponding to the given event source ID.
+    ///
+    /// This method does not reach out to the plugin-registry service's RPC
+    /// interface. Instead, it attempts to resolve the request against a local
+    /// cache in memory. Upon cache miss, this method attempts to enqueue a
+    /// cache update and returns immediately.
+    ///
+    /// # Arguments
+    ///
+    /// * event_source_id - the event source to retrieve generator IDs for
+    ///
+    /// # Returns
+    ///
+    /// * Ok(Some(generator_ids)) - A vector of generator IDs if there is an
+    ///   entry present in the cache.
+    ///
+    /// * Ok(None) - No entry corresponding to the given event_source_id was
+    ///   found in the cache, but an update request was successfully enqueued.
+    ///
+    /// * Err(e) - No entry corresponding to the given event_source_id was
+    ///   found in the cache, and we failed to enqueue an update request.
+    ///   Failures take two forms:
+    ///
+    ///   - GeneratorIdsCacheError::Retryable - e.g. we failed to enqueue an
+    ///     update request but it should work eventually if retried.
+    ///
+    ///   - GeneratorIdsCacheError::Fatal - something happened which has
+    ///     "poisoned" the GeneratorIdsCache instance such that all future
+    ///     calls to this method will fail.
+    pub async fn generator_ids_for_event_source(
+        &mut self,
+        event_source_id: Uuid,
+    ) -> Result<Option<Vec<Uuid>>, GeneratorIdsCacheError> {
+        if let Some(generator_ids) = self.generator_ids_cache.get(&event_source_id) {
+            Ok(Some(generator_ids)) // cache hit
+        } else if let Err(e) = self.updater_tx.try_send(event_source_id) {
+            Err(e.into()) // cache miss, failed to enqueue an update
+        } else {
+            Ok(None) // cache miss, successfully enqueued update
```

Comment:
not a must do, but: might be better suited as an enum { Hit, MissWithEnqueue, MissWithFail }

Reply:
It really does need to be a `Result<T, E>` where `E` (the miss with failure case) is actually an error. So regarding `T` I think `Option<Vec<Uuid>>` describes the shape of the success case pretty thoroughly.. it's either a list of IDs or it isn't.